### DISCUSSION TITLE
How can i type register function passing him to my own custom components?

### DISCUSSION THREAD

**Discussion by vitormarkis** [react-hook-form/react-hook-form#9917] [Q&A] [answered]
```ts
export interface TagFormFields {
  tag_name: string
  category: string
}
```
---
this code below is in the same file as left side image
```tsx
const {
    register,
    handleSubmit,
    reset,
    formState: { isSubmitSuccessful },
  } = useForm<TagFormFields>()
```

left side: form
right side: my own select component
![image](https://user-images.githubusercontent.com/121525239/218173305-775fee39-6940-4d93-b060-e54f4412abcb.png)

In the left side, i have a form, where i'm using a custom select input. To make the form works, i'm passing the register function to the select custom component, but how should i type the register properly to make the attribute `register: UseFormRegister<any>` receives  TagFormFields  automatically, and `field: string` receives keyof TagFormFields?

This type `TagFormFields` is already on the form environment, as i set this when i used useForm hook

**Comment by craigmcc** [comment]
For the record, I've run into the same thing.  Typing it as "UseFormRegister<any>" is the only thing I was able to come up with, either.

**Comment by Sajarin-M** [comment, marked as answer]
You can use `TypeScript Generics` and the `FieldPath` and `FieldValues` utility types in the React Hook Form as follows:
```js
import { FieldPath, FieldValues, UseFormRegister } from 'react-hook-form';

interface SelectProps<T extends FieldValues> {
  regiser: UseFormRegister<T>;
  field: FieldPath<T>;
}

function Select<T extends FieldValues>({ regiser, field }: SelectProps<T>) {
  return (
   <select {...regiser(field)} >
    {/* options */}
   </select>
  );
}
```

**Reply by vitormarkis** [reply]
It worked! thank you


```tsx
import { HTMLAttributes } from "react"
import FormFieldBox from "./FormFieldBox"

import { FieldPath, FieldValues, UseFormRegister } from "react-hook-form"

interface Props<T extends FieldValues> extends HTMLAttributes<HTMLInputElement> {
  parentAttrs?: HTMLAttributes<HTMLDivElement>
  field: FieldPath<T>
  register: UseFormRegister<T>
}

function Input<T extends FieldValues>(props: Props<T>){
  const { parentAttrs, field, register, ...rest } = props
  return (
    <FormFieldBox parentAttrs={parentAttrs}>
      <input
        {...register(field)}
        {...rest}
        type="text"
        className="w-full"
      />

    </FormFieldBox>
  )
}

export default Input
```
